infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/receive/PersistedWriterMetadata.java

```diff
@@ -0,0 +1,268 @@
+package org.corfudb.infrastructure.logreplication.receive;
+
+import lombok.Getter;
+import lombok.extern.slf4j.Slf4j;
+import org.corfudb.infrastructure.logreplication.proto.LogReplicationWriterMetadata.LogReplicationMetadataKey;
+import org.corfudb.infrastructure.logreplication.proto.LogReplicationWriterMetadata.LogReplicationMetadataVal;
+import org.corfudb.protocols.wireprotocol.logreplication.LogReplicationEntry;
+import org.corfudb.runtime.CorfuRuntime;
+import org.corfudb.runtime.CorfuStoreMetadata;
+import org.corfudb.runtime.collections.CorfuRecord;
+import org.corfudb.runtime.collections.CorfuStore;
+import org.corfudb.runtime.collections.Table;
+import org.corfudb.runtime.collections.TableOptions;
+import org.corfudb.runtime.collections.TxBuilder;
+import org.corfudb.runtime.view.Address;
+import java.util.UUID;
+
+/**
+ * The table persisted at the replication writer side.
+ * It records the logreader cluster's snapshot timestamp  and last log entry's timestamp, it has received and processed.
+ */
+@Slf4j
+public class PersistedWriterMetadata {
+    private static final String namespace = "CORFU_SYSTEM";
+    private static final String TABLE_PREFIX_NAME = "CORFU-REPLICATION-WRITER-";
+    String metadataTableName;
+
+    private CorfuStore corfuStore;
+
+    Table<LogReplicationMetadataKey, LogReplicationMetadataVal, LogReplicationMetadataVal> metadataTable;
+
+    CorfuRuntime runtime;
+
+    public PersistedWriterMetadata(CorfuRuntime rt, long siteConfigID, UUID primary, UUID dst) {
+        this.runtime = rt;
+        this.corfuStore = new CorfuStore(runtime);
+        metadataTableName = getPersistedWriterMetadataTableName(primary, dst);
+        try {
+            metadataTable = this.corfuStore.openTable(namespace,
+                    metadataTableName,
+                    LogReplicationMetadataKey.class,
+                    LogReplicationMetadataVal.class,
+                    null,
+                    TableOptions.builder().build());
+        } catch (Exception e) {
+            log.error("Caught an exception while open the table");
+            throw new ReplicationWriterException(e);
+        }
+        setupSiteConfigID(siteConfigID);
+    }
+
+    CorfuStoreMetadata.Timestamp getTimestamp() {
+        return corfuStore.getTimestamp();
+    }
+
+    TxBuilder getTxBuilder() {
+        return corfuStore.tx(namespace);
+    }
+
+    long query(CorfuStoreMetadata.Timestamp timestamp, PersistedWriterMetadataType key) {
```

Comment:
comments and also the boundary value -1 needs to be documented and maybe it should be a  constant.

Reply:
You could  also look into Optional class in java. It can make return values less prone to inducing errors.